File: java/src/main/java/za/co/entelect/challenge/Bot.java
package za.co.entelect.challenge;

import za.co.entelect.challenge.command.*;
import za.co.entelect.challenge.entities.*;
import za.co.entelect.challenge.enums.CellType;
import za.co.entelect.challenge.enums.Direction;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;

public class Bot {

    private Random random;
    private GameState gameState;
    private Opponent opponent;
    private MyWorm currentWorm;

    public Bot() {
        random = new Random(System.nanoTime());
    }

    public Command run(GameState gameState) {
        loadState(gameState);

        Worm enemyWorm = getFirstWormInRange();
        if (enemyWorm != null) {
            Direction direction = resolveDirection(currentWorm.position, enemyWorm.position);
            return new ShootCommand(direction);
        }

        System.out.println(currentWorm.position.x + " " + currentWorm.position.y);
        List<Cell> surroundingBlocks = getSurroundingCells(currentWorm.position.x, currentWorm.position.y);
        int cellIdx = random.nextInt(surroundingBlocks.size());

        Cell block = surroundingBlocks.get(cellIdx);
        if (block.type == CellType.AIR) {
            return new MoveCommand(block.x, block.y);
        } else if (block.type == CellType.DIRT) {
            return new DigCommand(block.x, block.y);
        }

        return new DoNothingCommand();
    }

    private void loadState(GameState gameState) {
        this.gameState = gameState;
        this.opponent = gameState.opponents[0];
        this.currentWorm = Arrays.stream(gameState.myPlayer.worms)
                .filter(myWorm -> myWorm.id == gameState.currentWormId)
                .findFirst()
                .get();
    }

    private Worm getFirstWormInRange() {
        for (Worm enemyWorm : opponent.worms) {
            if (euclideanDistance(currentWorm.position, enemyWorm.position) <= currentWorm.weapon.range) {
                return enemyWorm;
            }
        }

        return null;
    }

    private List<Cell> getSurroundingCells(int x, int y) {
        ArrayList<Cell> cells = new ArrayList<>();
        for (int i = x - 1; i <= x + 1; i++) {
            for (int j = y - 1; j <= y + 1; j++) {
                // Don't include the current position
                if (i != x && j != y && isValidCoordinate(i, j)) {
                    cells.add(gameState.map[j][i]);
                }
            }
        }

        return cells;
    }

    private int euclideanDistance(Position a, Position b) {
        return (int) (Math.sqrt(Math.pow(a.x - b.x, 2) + Math.pow(a.y - b.y, 2)));
    }

    private boolean isValidCoordinate(int i, int j) {

        if (i < 0 || i >= gameState.mapSize) {
            return false;
        }

        if (j < 0 || j >= gameState.mapSize) {
            return false;
        }

        return true;
    }

    private Direction resolveDirection(Position a, Position b) {
        StringBuilder builder = new StringBuilder();

        int verticalComponent = b.y - a.y;
        int horizontalComponent = b.x - a.x;

        if (verticalComponent < 0) {
            builder.append('N');
        } else if (verticalComponent > 0) {
            builder.append('S');
        }

        if (horizontalComponent < 0) {
            builder.append('W');
        } else if (horizontalComponent > 0) {
            builder.append('E');
        }

        return Direction.valueOf(builder.toString());
    }
}

File: starter-bots/scala/src/main/scala/za/co/entelect/challenge/Bot.scala
package za.co.entelect.challenge

import za.co.entelect.challenge.commands.{Command, DigCommand, DoNothingCommand, MoveCommand, ShootCommand}
import za.co.entelect.challenge.entities.{Cell, GameState, MyWorm, Opponent, Position, Worm}
import za.co.entelect.challenge.enums.CellType.CellTypes
import za.co.entelect.challenge.enums.{CellType, Direction}
import za.co.entelect.challenge.enums.Direction.{Direction, DirectionObject, Directions}

import scala.collection.mutable.ListBuffer
import scala.util.Random

class Bot(gameState: GameState, random: Random){
  private val opponent: Opponent  = gameState.opponents.head
  private val currentWorm: MyWorm = getCurrentWorm(gameState)

  private def euclideanDistance(aX: Int, aY: Int, bX: Int, bY: Int): Int = {
    Math.sqrt(
      Math.pow(aX - bX, 2) + Math.pow(aY - bY, 2)
    ).toInt
  }

  private def isValidCoordinate(x: Int, y: Int): Boolean = {
    x >= 0 && x < gameState.mapSize && y >= 0 && y < gameState.mapSize
  }

  private def getCurrentWorm(gameState: GameState): MyWorm = {
    gameState.myPlayer.worms
      .filter(myWorm => myWorm.id == gameState.currentWormId)
      .head
  }

  private def resolveDirection(a: Position, b: Position): Direction = {
    val verticalComponent = b.y - a.y
    val horizontalComponent = b.x - a.x

    val builder: StringBuilder = new StringBuilder()

    if     (verticalComponent < 0) builder.append('N')
    else if(verticalComponent > 0) builder.append('S')

    if     (horizontalComponent < 0) builder.append('W')
    else if(horizontalComponent > 0) builder.append('E')

    Directions.withNameInsensitive(builder.toString())
  }

  private def getSurroundingCells(x: Int, y: Int): List[Cell] = {
    val cells: ListBuffer[Cell] = ListBuffer.empty[Cell]

    for (
      i <- Range(x - 1, x + 2);
      j <- Range(y - 1, y + 2)
    ) {
        if(i != x && j != y && isValidCoordinate(i, j)) {
          cells += gameState.map(j)(i)
        }
      }

    cells.toList
  }

  private def constructFireDirectionLines(range: Int): List[List[Cell]] = {
    val directionLines: ListBuffer[List[Cell]] = ListBuffer.empty[List[Cell]]

    for (direction: Direction <- Directions.values) {
      val directionLine: ListBuffer[Cell] = ListBuffer.empty[Cell]
      val directionObject: DirectionObject = Direction.matchDirection(direction)

      for (directionMultiplier <- Range(1, range + 1)) {
        val coordinateX: Int = currentWorm.position.x + (directionMultiplier * directionObject.x)
        val coordinateY: Int = currentWorm.position.y + (directionMultiplier * directionObject.y)

        val checkValidCoordinate           = isValidCoordinate(coordinateX, coordinateY)
        val checkOutOfRange                = !(euclideanDistance(currentWorm.position.x, currentWorm.position.y, coordinateX, coordinateY) > range)
        val checkCellType: Cell => Boolean = { cell => cell.tpe == CellTypes.AIR }

        if (checkValidCoordinate && checkOutOfRange) {
          val cell: Cell = gameState.map(coordinateY)(coordinateX)
          if (checkCellType(cell)) {
            directionLine += gameState.map(coordinateY)(coordinateX)

          }
        }
      }
      directionLines += directionLine.toList
    }
    directionLines.toList
  }

  private def getFirstWormInRange(): Option[Worm] = {
    val cells: List[String] =
      for {
        cells <- constructFireDirectionLines(currentWorm.weapon.range)
        cell  <- cells
      } yield s"${cell.x}_${cell.y}"

    opponent.worms.find(enemyWorm => {
      val enemyPosition = s"${enemyWorm.position.x}_${enemyWorm.position.y}"
      cells.contains(enemyPosition)
    })
  }

  def run(): Command = {
    val enemyWorm = getFirstWormInRange()
    enemyWorm match {
      case Some(worm) => {
        val direction: Direction = resolveDirection(currentWorm.position, worm.position)
        ShootCommand(direction)
      }
      case None       => {
        val surroundingBlocks: List[Cell] = getSurroundingCells(currentWorm.position.x, currentWorm.position.y)
        println(s"\n =========== \n ${surroundingBlocks.size} \n =========== \n")
        val cellIdx = random.nextInt(surroundingBlocks.length)
        println(s"\n =========== \n ${cellIdx} \n =========== \n")

        val block: Cell = surroundingBlocks(cellIdx)
        block.tpe match {
          case CellTypes.AIR        => MoveCommand(block.x, block.y)
          case CellTypes.DIRT       => DigCommand(block.x, block.y)
          case CellTypes.DEEP_SPACE => DoNothingCommand()
        }
      }
    }
  }
}

object Bot {
  def apply(gameState: GameState, random: Random): Bot = {
    new Bot(gameState, random)
  }
}

